src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java

```diff
@@ -264,11 +308,34 @@ private SpoolMessage getSpoolMessageFromRs(long messageId, ResultSet rs) throws
         }
     }
 
+    /**
+     * This method checks for connection exists or not if not creates connection.
+     * @throws SQLException When fails to get Database Connection
+     */
+    synchronized void ensureConnection() throws SQLException {
+        if (dbConnection == null || dbConnection.isClosed()) {
+            try (LockScope ignored = LockScope.lock(dbConnectionLock.writeLock())) {
+                init();
```

Comment:
so `ensureConnection` is always called within a read lock, then init acquires a write lock.   would this deadlock?

Reply:
from ReentrantReadWriteLock: 
```
However, upgrading from a read lock to the write lock is <b>not</b> possible.
```